File: src/algorithms/token-bucket.ts
import { Store } from '../stores/store.interface';
import { BucketState, RateLimiterConfig } from '../types';

export class TokenBucket {
  private capacity: number;
  private refillRate: number;

  constructor(
    private store: Store,
    private key: string,
    config: RateLimiterConfig
  ) {
    this.capacity = config.capacity;
    this.refillRate = config.refillRate;
  }

  async isAllowed(tokens: number = 1): Promise<boolean> {
    const state = await this.getOrCreateState();
    const now = Date.now();
    
    // Calculate tokens added since last refill
    const timePassed = (now - state.lastRefill) / 1000; // Convert to seconds
    const tokensToAdd = timePassed * this.refillRate;
    
    let currentTokens = Math.min(
      state.tokens + tokensToAdd,
      this.capacity
    );

    if (currentTokens >= tokens) {
      currentTokens -= tokens;
      await this.store.set(this.key, {
        tokens: currentTokens,
        lastRefill: now
      });
      return true;
    }

    return false;
  }

  async consume(tokens: number = 1): Promise<number> {
    const state = await this.getOrCreateState();
    const now = Date.now();

    const timePassed = (now - state.lastRefill) / 1000;
    const tokensToAdd = timePassed * this.refillRate;

    let currentTokens = Math.min(
      state.tokens + tokensToAdd,
      this.capacity
    );

    currentTokens = Math.max(0, currentTokens - tokens);

    await this.store.set(this.key, {
      tokens: currentTokens,
      lastRefill: now
    });

    return currentTokens;
  }

  async getStatus() {
    const state = await this.getOrCreateState();
    const now = Date.now();

    const timePassed = (now - state.lastRefill) / 1000;
    const tokensToAdd = timePassed * this.refillRate;

    const currentTokens = Math.min(
      state.tokens + tokensToAdd,
      this.capacity
    );

    return {
      tokens: currentTokens,
      lastRefill: state.lastRefill,
      capacity: this.capacity,
      refillRate: this.refillRate
    };
  }

  async reset(): Promise<void> {
    await this.store.delete(this.key);
  }

  private async getOrCreateState(): Promise<BucketState> {
    let state = await this.store.get(this.key);
    if (!state) {
      state = {
        tokens: this.capacity,
        lastRefill: Date.now()
      };
      await this.store.set(this.key, state);
    }
    return state;
  }
}

File: app/algorithms/token_bucket.py
"""Token Bucket Algorithm (In-Memory Version for Testing)"""

import time


class TokenBucket:
    """
    Token Bucket rate limiter (in-memory, for testing).
    
    Production deployments should use the Redis-backed version in services.rate_limiter.
    """
    
    def __init__(self, capacity, refill_rate):
        """
        Args:
            capacity (float): Maximum tokens in bucket
            refill_rate (float): Tokens generated per second
        """
        self.capacity = capacity
        self.refill_rate = refill_rate
        self.tokens = capacity
        self.last_refill = time.time()
    
    def allow_request(self, current_time=None):
        """
        Check if request is allowed
        
        Args:
            current_time (float): Current time (default: now)
        
        Returns:
            dict: {allowed: bool, remaining_tokens: int}
        """
        if current_time is None:
            current_time = time.time()
        
        # Calculate elapsed time
        elapsed = current_time - self.last_refill
        
        # Refill tokens
        tokens_to_add = elapsed * self.refill_rate
        self.tokens = min(self.tokens + tokens_to_add, self.capacity)
        
        # Update last refill
        self.last_refill = current_time
        
        # Check if allowed
        allowed = self.tokens >= 1.0
        
        if allowed:
            self.tokens -= 1.0
        
        return {
            "allowed": allowed,
            "remaining_tokens": max(0, int(self.tokens))
        }
